File: backend-go/internal/modules/system/service/menu_service.go
package service

import (
	"drip-admin/backend-go/internal/common"
	"github.com/gin-gonic/gin"
)

func (s *Server) Menus(c *gin.Context) {
	rows, err := s.allMenuTree()
	successOrError(c, rows, err)
}

func (s *Server) CreateMenu(c *gin.Context) {
	var request MenuSaveRequest
	if err := common.BindJSON(c, &request); err != nil {
		common.HandleError(c, err)
		return
	}
	id, err := s.saveMenu(0, request, true)
	successOrError(c, id, err)
}

func (s *Server) UpdateMenu(c *gin.Context) {
	id, err := idParam(c)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	var request MenuSaveRequest
	if err := common.BindJSON(c, &request); err != nil {
		common.HandleError(c, err)
		return
	}
	_, err = s.saveMenu(id, request, false)
	successOrError(c, nil, err)
}

func (s *Server) DeleteMenu(c *gin.Context) {
	id, err := idParam(c)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	var count int64
	if err := s.db.Model(&SysMenu{}).Where("parent_id = ? and deleted = 0", id.Int64()).Count(&count).Error; err != nil {
		common.HandleError(c, err)
		return
	}
	if count > 0 {
		common.HandleError(c, common.NewBusinessError(400301, "operation failed"))
		return
	}
	successOrError(c, nil, s.db.Model(&SysMenu{}).Where("id = ?", id.Int64()).Update("deleted", 1).Error)
}

func (s *Server) MenuStatus(c *gin.Context) {
	id, err := idParam(c)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	status, err := statusRequest(c)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	successOrError(c, nil, s.db.Model(&SysMenu{}).Where("id = ? and deleted = 0", id.Int64()).Update("status", status).Error)
}

func (s *Server) saveMenu(id common.Int64String, request MenuSaveRequest, create bool) (common.Int64String, error) {
	if err := common.RequiredString(request.Name, "name"); err != nil {
		return 0, err
	}
	if err := common.RequiredString(request.Type, "type"); err != nil {
		return 0, err
	}
	parentID := idOrZero(request.ParentID)
	if parentID != 0 {
		var count int64
		if err := s.db.Model(&SysMenu{}).Where("id = ? and deleted = 0", parentID.Int64()).Count(&count).Error; err != nil {
			return 0, err
		}
		if count == 0 {
			return 0, common.NotFound("operation failed")
		}
	}
	if !create && (parentID == id || s.isMenuDescendant(id, parentID)) {
		return 0, common.NewBusinessError(400000, "operation failed")
	}
	values := map[string]any{
		"parent_id":       parentID,
		"name":            request.Name,
		"type":            request.Type,
		"path":            optionalString(request.Path),
		"component":       optionalString(request.Component),
		"permission_code": optionalString(request.PermissionCode),
		"icon":            optionalString(request.Icon),
		"sort":            intOrDefault(request.Sort, 0),
		"visible":         intOrDefault(request.Visible, 1),
		"status":          intOrDefault(request.Status, 1),
	}
	if create {
		row := SysMenu{
			ID:             common.NewID(),
			ParentID:       parentID,
			Name:           request.Name,
			Type:           request.Type,
			Path:           optionalString(request.Path),
			Component:      optionalString(request.Component),
			PermissionCode: optionalString(request.PermissionCode),
			Icon:           optionalString(request.Icon),
			Sort:           intOrDefault(request.Sort, 0),
			Visible:        intOrDefault(request.Visible, 1),
			Status:         intOrDefault(request.Status, 1),
		}
		return row.ID, s.db.Create(&row).Error
	}
	return id, s.db.Model(&SysMenu{}).Where("id = ? and deleted = 0", id.Int64()).Updates(values).Error
}

func (s *Server) isMenuDescendant(id common.Int64String, maybeChild common.Int64String) bool {
	if maybeChild == 0 {
		return false
	}
	var rows []SysMenu
	_ = s.db.Select("id", "parent_id").Where("deleted = 0").Find(&rows).Error
	children := map[int64][]common.Int64String{}
	for _, row := range rows {
		children[row.ParentID.Int64()] = append(children[row.ParentID.Int64()], row.ID)
	}
	var walk func(common.Int64String) bool
	walk = func(current common.Int64String) bool {
		for _, child := range children[current.Int64()] {
			if child == maybeChild || walk(child) {
				return true
			}
		}
		return false
	}
	return walk(id)
}

func (s *Server) allMenuTree() ([]MenuTreeVo, error) {
	var rows []SysMenu
	if err := s.db.Where("deleted = 0").Order("sort asc, id asc").Find(&rows).Error; err != nil {
		return nil, err
	}
	return buildMenuTree(rows, false), nil
}

func (s *Server) menuTreeForUser(userID common.Int64String) ([]MenuTreeVo, error) {
	roles, err := s.roleCodes(userID)
	if err != nil {
		return nil, err
	}
	var rows []SysMenu
	if containsString(roles, "SUPER_ADMIN") {
		if err := s.db.Where("status = 1 and deleted = 0").Order("sort asc, id asc").Find(&rows).Error; err != nil {
			return nil, err
		}
		return buildMenuTree(rows, true), nil
	}
	var allRows []SysMenu
	if err := s.db.Where("status = 1 and deleted = 0").Order("sort asc, id asc").Find(&allRows).Error; err != nil {
		return nil, err
	}
	var directRows []SysRoleMenu
	if err := s.db.Table("sys_role_menu").
		Select("distinct sys_role_menu.*").
		Joins("join sys_user_role on sys_user_role.role_id = sys_role_menu.role_id").
		Where("sys_user_role.user_id = ?", userID.Int64()).
		Find(&directRows).Error; err != nil {
		return nil, err
	}
	byID := map[int64]SysMenu{}
	for _, row := range allRows {
		byID[row.ID.Int64()] = row
	}
	visible := map[int64]bool{}
	for _, direct := range directRows {
		current := direct.MenuID
		for current != 0 {
			menu, ok := byID[current.Int64()]
			if !ok || visible[current.Int64()] {
				break
			}
			visible[current.Int64()] = true
			current = menu.ParentID
		}
	}
	rows = make([]SysMenu, 0, len(allRows))
	for _, row := range allRows {
		if visible[row.ID.Int64()] {
			rows = append(rows, row)
		}
	}
	return buildMenuTree(rows, true), nil
}

func buildMenuTree(rows []SysMenu, excludeButton bool) []MenuTreeVo {
	byID := map[int64]*MenuTreeVo{}
	order := make([]common.Int64String, 0, len(rows))
	for _, row := range rows {
		if excludeButton && row.Type == "BUTTON" {
			continue
		}
		vo := MenuTreeVo{ID: row.ID, ParentID: row.ParentID, Name: row.Name, Type: row.Type, Path: row.Path, Component: row.Component, PermissionCode: row.PermissionCode, Icon: row.Icon, Sort: row.Sort, Visible: row.Visible, Status: row.Status, Children: []MenuTreeVo{}}
		byID[row.ID.Int64()] = &vo
		order = append(order, row.ID)
	}
	for _, id := range order {
		row := byID[id.Int64()]
		parent := byID[row.ParentID.Int64()]
		if row.ParentID != 0 && parent != nil {
			parent.Children = append(parent.Children, *row)
		}
	}
	roots := []MenuTreeVo{}
	for _, id := range order {
		row := byID[id.Int64()]
		if row.ParentID == 0 || byID[row.ParentID.Int64()] == nil {
			roots = append(roots, *row)
		}
	}
	return roots
}

File: backend-python/app/modules/system/service/menu_service.py
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.common.errors import BusinessError, bad_request, not_found
from app.common.id import new_id
from app.modules.system.dto.menu_request import MenuSaveRequest
from app.modules.system.entity import SysMenu, SysRole, SysRoleMenu, SysUserRole
from app.modules.system.service.permission_service import PermissionService


class MenuService:
    def __init__(self, db: AsyncSession) -> None:
        self.db = db
        self.permissions = PermissionService(db)

    async def menu_tree_for_user(self, user_id: int) -> list[dict]:
        roles = await self.permissions.role_codes(user_id)
        if "SUPER_ADMIN" in roles:
            query = select(SysMenu).where(SysMenu.status == 1, SysMenu.deleted == 0)
        else:
            query = (
                select(SysMenu)
                .join(SysRoleMenu, SysRoleMenu.menu_id == SysMenu.id)
                .join(SysRole, SysRole.id == SysRoleMenu.role_id)
                .join(SysUserRole, SysUserRole.role_id == SysRole.id)
                .where(
                    SysUserRole.user_id == user_id,
                    SysRole.status == 1,
                    SysRole.deleted == 0,
                    SysMenu.status == 1,
                    SysMenu.deleted == 0,
                )
            )
        rows = list((await self.db.scalars(query)).unique().all())
        return self._tree(rows)

    async def all_tree(self) -> list[dict]:
        rows = list(
            (
                await self.db.scalars(
                    select(SysMenu)
                    .where(SysMenu.deleted == 0)
                    .order_by(SysMenu.sort, SysMenu.id)
                )
            ).all()
        )
        return self._tree(rows)

    async def save(self, menu_id: int | None, request: MenuSaveRequest) -> str | None:
        parent_id = request.parent_id or 0
        await self._assert_valid_parent(menu_id, parent_id)
        permission_code = _optional(request.permission_code)
        if permission_code:
            duplicate = select(SysMenu.id).where(
                SysMenu.permission_code == permission_code,
                SysMenu.deleted == 0,
            )
            if menu_id is not None:
                duplicate = duplicate.where(SysMenu.id != menu_id)
            if await self.db.scalar(duplicate) is not None:
                raise BusinessError(409000, "permissionCode already exists")
        values = {
            "parent_id": parent_id,
            "name": request.name.strip(),
            "type": request.type.strip(),
            "path": _optional(request.path),
            "component": _optional(request.component),
            "permission_code": permission_code,
            "icon": _optional(request.icon),
            "sort": 0 if request.sort is None else request.sort,
            "visible": 1 if request.visible is None else request.visible,
            "status": 1 if request.status is None else request.status,
        }
        if menu_id is None:
            row = SysMenu(id=new_id(), deleted=0, **values)
            self.db.add(row)
            await self.db.commit()
            return str(row.id)
        row = await self._find(menu_id)
        for key, value in values.items():
            setattr(row, key, value)
        await self.db.commit()
        return None

    async def delete(self, menu_id: int) -> None:
        row = await self._find(menu_id)
        child_count = await self.db.scalar(
            select(func.count()).select_from(SysMenu).where(
                SysMenu.parent_id == menu_id,
                SysMenu.deleted == 0,
            )
        )
        if child_count:
            raise BusinessError(409000, "menu contains child nodes")
        row.deleted = 1
        await self.db.commit()

    async def update_status(self, menu_id: int, status: int | None) -> None:
        row = await self._find(menu_id)
        row.status = 1 if status is None else status
        await self.db.commit()

    async def _assert_valid_parent(self, menu_id: int | None, parent_id: int) -> None:
        if parent_id == 0:
            return
        await self._find(parent_id)
        if menu_id is None:
            return
        if parent_id == menu_id:
            raise bad_request("operation failed")
        rows = (await self.db.execute(select(SysMenu.id, SysMenu.parent_id).where(SysMenu.deleted == 0))).all()
        descendants = _descendants(menu_id, rows)
        if parent_id in descendants:
            raise bad_request("operation failed")

    async def _find(self, menu_id: int) -> SysMenu:
        row = await self.db.scalar(select(SysMenu).where(SysMenu.id == menu_id, SysMenu.deleted == 0))
        if row is None:
            raise not_found()
        return row

    def _tree(self, rows: list[SysMenu]) -> list[dict]:
        ordered = sorted(rows, key=lambda row: (row.sort, row.id))
        nodes = {row.id: self._node(row) for row in ordered}
        roots: list[dict] = []
        for row in ordered:
            node = nodes[row.id]
            parent = nodes.get(row.parent_id)
            if parent is None:
                roots.append(node)
            else:
                parent["children"].append(node)
        return roots

    @staticmethod
    def _node(row: SysMenu) -> dict:
        return {
            "id": str(row.id),
            "parentId": str(row.parent_id),
            "name": row.name,
            "type": row.type,
            "path": row.path,
            "component": row.component,
            "permissionCode": row.permission_code,
            "icon": row.icon,
            "sort": row.sort,
            "visible": row.visible,
            "status": row.status,
            "children": [],
        }


def _descendants(parent_id: int, rows: list[tuple[int, int]]) -> set[int]:
    children: dict[int, list[int]] = {}
    for child_id, current_parent_id in rows:
        children.setdefault(current_parent_id, []).append(child_id)
    result: set[int] = set()
    stack = list(children.get(parent_id, []))
    while stack:
        current = stack.pop()
        if current in result:
            continue
        result.add(current)
        stack.extend(children.get(current, []))
    return result


def _optional(value: str | None) -> str | None:
    return value.strip() if value and value.strip() else None
